code/datums/elements/immerse.dm

```diff
@@ -0,0 +1,387 @@
+	SIGNAL_HANDLER
+	RegisterSignals(source, list(COMSIG_ATOM_ABSTRACT_ENTERED, COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZED_ON), PROC_REF(on_init_or_entered))
+	RegisterSignal(source, COMSIG_ATOM_ABSTRACT_EXITED, PROC_REF(on_atom_exited))
+	attached_turfs_and_movables += source
+	for(var/atom/movable/movable as anything in source)
+		on_init_or_entered(source, movable)
+
+///Stops the element from affecting on the turf and its contents. Called on Detach() or when TRAIT_IMMERSE_STOPPED is added.
+/datum/element/immerse/proc/stop_immersion(turf/source)
+	SIGNAL_HANDLER
+	UnregisterSignal(source, list(COMSIG_ATOM_ABSTRACT_ENTERED, COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZED_ON, COMSIG_ATOM_ABSTRACT_EXITED))
+	for(var/atom/movable/movable as anything in attached_turfs_and_movables[source])
+		remove_from_element(source, movable)
+	attached_turfs_and_movables -= source
+
+/**
+ * If the movable is within the right layers and planes, not in the list of movable types to ignore,
+ * or already affected by the element for that matter, Signals will be registered and,
+ * unless the movable (or whatever it's buckled to) is flying, it'll appear as if immersed in that water.
+ */
+/datum/element/immerse/proc/on_init_or_entered(turf/source, atom/movable/movable)
+	SIGNAL_HANDLER
+	if(movable.layer >= ABOVE_ALL_MOB_LAYER || !ISINRANGE(movable.plane, MUTATE_PLANE(FLOOR_PLANE, source), MUTATE_PLANE(GAME_PLANE_UPPER_FOV_HIDDEN, source)))
+		return
+	if(HAS_TRAIT(movable, TRAIT_IMMERSED))
+		return
+	if(is_type_in_typecache(movable, movables_to_ignore))
+		return
+
+	var/atom/movable/buckled
+	if(isliving(movable))
+		var/mob/living/living_mob = movable
+		RegisterSignal(living_mob, COMSIG_LIVING_SET_BUCKLED, PROC_REF(on_set_buckled))
+		buckled = living_mob.buckled
+
+	try_immerse(movable, buckled)
+	RegisterSignal(movable, COMSIG_QDELETING, PROC_REF(on_movable_qdel))
+	LAZYADD(attached_turfs_and_movables[source], movable)
+	ADD_TRAIT(movable, TRAIT_IMMERSED, ELEMENT_TRAIT(src))
+
+/datum/element/immerse/proc/on_movable_qdel(atom/movable/source)
+	SIGNAL_HANDLER
+	remove_from_element(source.loc, source)
+
+/**
+ * The main proc, which adds a visual overlay to the movable that has entered the turf to make it look immersed.
+ * It's kind of iffy but basically, we want the overlay to cover as much area as needed to
+ * avoid the movable's icon from spilling horizontally or below.
+ * Also, while these visual overlays are mainly cached movables, for certain movables, such as living mobs,
+ * we want them to have their own unique vis overlay with additional signals registered.
+ * This allows the vis overlay to look more or less unchanged while its owner is spinning or resting
+ * without otherwise affecting other movables with identical overlays.
+ */
+/datum/element/immerse/proc/add_immerse_overlay(atom/movable/movable)
+	var/icon/movable_icon = icon(movable.icon)
+	var/width = movable_icon.Width() || world.icon_size
+	var/height = movable_icon.Height() || world.icon_size
```

Comment:
this makes me a tad uncomfortable. icon operations can be quite slow, and since this is I think off move() I'm somewhat concerned.
we've done similar in lighting code, so maybe? it's ok. idk

Reply:
Considering it's only done once when the movable is dipped into the water until it resurfaces, it's probably not that bad but I feel you.
Maybe, one day. we should cache icon width and height and use getter macros instead, trading 16 bytes or so per atom or movable for better performance, though we need to evaluate if that's even worth it. 🤔